### PR TITLE
Add support for publishing to Custom domains

### PR DESCRIPTION
**Pull request by matthewdavidrodgers**
With the release of Custom Domains for workers, users can publish directly to a custom domain on a route, rather than creating a dummy DNS record first and manually pointing the worker over - this adds the same support to wrangler.

Users declare routes as normal, but to indicate that a route should be treated as a custom domain, a user simply uses the object format in the toml file, but with a new key: custom_domain (i.e. `routes = [{ pattern = "api.example.com", custom_domain = true }]`)

When wrangler sees a route like this, it peels them off from the rest of the routes and publishes them separately, using the /domains api. This api is very defensive, erroring eagerly if there are conflicts in existing Custom Domains or managed DNS records. In the case of conflicts, wrangler prompts for confirmation, and then retries with parameters to indicate overriding is allowed.

### CODE REVIEW

File: packages/wrangler/src/publish.ts
Comment (matthewdavidrodgers) [inline packages/wrangler/src/publish.ts:41]: These type definitions look scary, but they simply extract the object version of a route definition in the config (i.e. `routes: [{ pattern: "api.example.com" }]` rather than `routes: ["api.example.com"]`). A custom domain must be defined in the object form (to provide the flag `custom_domain: true`), so this keeps us in sync whenever the Config definition changes

File: packages/wrangler/src/publish.ts
Comment (matthewdavidrodgers) [inline packages/wrangler/src/publish.ts:81]: Originally tried to use a regex here, but it was too permissive when there were multiple quote bounded substrings. This will always grab the first one (and even works when the string begins with a quote)

**Review by Electroid — COMMENTED** [review]
Maybe @threepointone would have a good opinion on this, but I wonder if it makes sense for this feature to be separate from routes, something like:

```toml
domains = ["api.example.com"]
```

The reason I propose this is because I think we want to push folks to use domains instead of routes, since it sets up DNS, SSL, and all that jazz. If we jam them together, it might get confusing.

**Comment by threepointone** [comment]
Another option is to use `{ domain = 'api.example.com' }` as the form. 

Thanks a lot for the PR @matthewdavidrodgers! This looks extensive, and thank you for adding tests! We'll review this soon.

File: packages/wrangler/src/publish.ts
Comment (threepointone) [inline packages/wrangler/src/publish.ts:53]:
```suggestion
    const isCustomDomain = Boolean("custom_domain" in route && route.custom_domain);
```

File: packages/wrangler/src/publish.ts
Comment (threepointone) [inline packages/wrangler/src/publish.ts:105]: This function deserves a comment, just explaning what it does

**Review by threepointone — COMMENTED** [review]
Love this PR. I honestly don't have any notes, but I'll let @petebacondarwin have a quick look at this before I stamp it. Could you also rebase to head? We've landed a few changes since this came in.

**Review by petebacondarwin — APPROVED** [review]
Also love this PR. Made a few comments but none blocking merging this.
Thanks @matthewdavidrodgers.

**Comment by Electroid** [comment]
I think @threepointone’s idea of `routes = [{domain = “api.example.com”]` is worth considering here. It’s compact and succinct than having a separate boolean. It’s also not really a “pattern”

File: packages/wrangler/src/publish.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/publish.ts:77]: so, went to the api team today to try and see what ways we can provide structured data in an error response, and there weren't a whole lot of options they felt would work. something simple and trivially parseable within the error message seems to be the best way to handle this.

currently, the error messages look something like this:
`Cannot create Custom Domain "<domain>": Custom Domain already exists and points to a different script; retry and use option "override_existing_origin" to override`
which is why I parse out the first quote bounded substring.

do we feel better if i change the error message to look something like this:
`"<domain>" could not be created as a Custom Domain; already exists and points to a different script; retry and use option "override_existing_origin" to override`
so at least the conflicting domains are the first thing in the message? maybe making the parsing less fraught?

File: packages/wrangler/src/publish.ts
Comment (matthewdavidrodgers) [reply packages/wrangler/src/publish.ts:81]: there's no current situation where this could occur - I authored the code that generates these messages - but I suppose all bets are off if the code drifts. I could add a comment reinforcing the general structure in the api

**Review by threepointone — APPROVED** [review]
Ok, I'm ready to land this!

**Pull request merged**
